// @substrate/smoldot-discovery@2.0.2/dist/commonjs/types/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueueFullError = exports.JsonRpcDisabledError = exports.CrashError = exports.AlreadyDestroyedError = exports.WellKnownChain = void 0;
/**
 * List of popular chains that are likely to be connected to.
 *
 * The values in this enum correspond to the `id` field of the relevant chain specification.
 */
exports.WellKnownChain = {
    polkadot: "polkadot",
    ksmcc3: "ksmcc3",
    rococo_v2_2: "rococo_v2_2",
    westend2: "westend2",
    paseo: "paseo",
};
class AlreadyDestroyedError extends Error {
    constructor() {
        super();
        this.name = "AlreadyDestroyedError";
    }
}
exports.AlreadyDestroyedError = AlreadyDestroyedError;
class CrashError extends Error {
    constructor(message) {
        super(message);
        this.name = "CrashError";
    }
}
exports.CrashError = CrashError;
class JsonRpcDisabledError extends Error {
    constructor() {
        super();
        this.name = "JsonRpcDisabledError";
    }
}
exports.JsonRpcDisabledError = JsonRpcDisabledError;
class QueueFullError extends Error {
    constructor() {
        super();
        this.name = "QueueFullError";
    }
}
exports.QueueFullError = QueueFullError;
